api/v1alpha1/k6_types.go

```diff
@@ -52,17 +53,18 @@ type K6Scuttle struct {
 
 // K6Spec defines the desired state of K6
 type K6Spec struct {
-	Script      K6Script               `json:"script"`
-	Parallelism int32                  `json:"parallelism"`
-	Separate    bool                   `json:"separate,omitempty"`
-	Arguments   string                 `json:"arguments,omitempty"`
-	Ports       []corev1.ContainerPort `json:"ports,omitempty"`
-	Starter     Pod                    `json:"starter,omitempty"`
-	Runner      Pod                    `json:"runner,omitempty"`
-	Quiet       string                 `json:"quiet,omitempty"`
-	Paused      string                 `json:"paused,omitempty"`
-	Scuttle     K6Scuttle              `json:"scuttle,omitempty"`
-	Cleanup     Cleanup                `json:"cleanup,omitempty"`
+	Script           K6Script                      `json:"script"`
+	Parallelism      int32                         `json:"parallelism"`
+	Separate         bool                          `json:"separate,omitempty"`
+	Arguments        string                        `json:"arguments,omitempty"`
+	Ports            []corev1.ContainerPort        `json:"ports,omitempty"`
+	Starter          Pod                           `json:"starter,omitempty"`
+	Runner           Pod                           `json:"runner,omitempty"`
+	Quiet            string                        `json:"quiet,omitempty"`
+	Paused           string                        `json:"paused,omitempty"`
+	Scuttle          K6Scuttle                     `json:"scuttle,omitempty"`
+	Cleanup          Cleanup                       `json:"cleanup,omitempty"`
+	ImagePullSecrets []corev1.LocalObjectReference `json:"imagePullSecrets,omitempty"`
```

Comment:
@kneemaa thanks for opening the PR! I haven't time to run it yet but just looking briefly around, why do you have `ImagePullSecrets` in the spec? It's already in the pod - why duplicates?

Reply:
@yorugac I took it out just now.

Probably was left over from my troubleshooting. Im still unable to figure out what I'm missing to pass from the K6Spec to the actual pod Kubernetes creates. (`make manifests` doesn't yield any changes so I'm guessing on where to set my PullSecret)

The last thing I tried was having the Initializer log out what value it had for `ImagePullSecrets` and it's only come back as `[{}]`.

Is there something basic I'm missing?

```yaml
# https://github.com/grafana/k6-operator/blob/main/README.md#parallelism
apiVersion: k6.io/v1alpha1
kind: K6
metadata:
  name: k6-apache-stress
  namespace: chaos
spec:
  parallelism: 10
  arguments: --out statsd
  script:
    configMap:
      name: apache-stress-test
      file: test.js
  starter:
    serviceAccountName: k6-operator-controller
    image: <private registry>/k6-test-nema:latest
    imagePullSecrets:
      - name: <secret name>
  runner:
    image: <private registry>/k6-test-runner:latest
    imagePullSecrets:
      - name: <secret name>
    serviceAccountName: k6-operator-controller
```

Reply:
Hi @kneemaa, I've finally gotten to testing this PR, my apologies for the delay.
Your PR worked fine for me, I tested it with this setup:
```
  runner:
    image: ...
    imagePullSecrets:
      - name: ...
```
and with docker hub private repo. (it _shouldn't_ make a difference but noting just in case)

So not sure what the problem is in your test... If initializer and other pods started, then private image was pulled successfully, right? Kubernetes should show it explicitly in resource description. Could you please clarify what exactly didn't work then?

Reply:
Hi @yorugac,

I wonder if I'm missing something in my local setup. My steps are...

1. My operator is running with image generated by `Dockerfile.controller`
2. Generate updated manifests K8 manifests
`kustomize build config/default > k6-operator-controller-manager.yml`
3. Apply updated CRD
4. Apply K6 Resource with only the Starter using the private image
```yaml
apiVersion: k6.io/v1alpha1
kind: K6
metadata:
  name: k6-apache-stress
  namespace: chaos
spec:
  parallelism: 2
  arguments: --out statsd
  script:
    configMap:
      name: apache-stress-test
      file: test.js
  starter:
    serviceAccountName: k6-operator-controller
    image: <repo>/operator:controller-v0.0.7
    imagePullSecrets:
      - name: stash-artifactory
  runner:
    # image: <repo>/operator:controller-v0.0.7
    serviceAccountName: k6-operator-controller
    env:
      - name: K6_STATSD_ENABLE_TAGS
        value: "true"
      - name: K6_STATSD_ADDR
        value: datadog.datadog:8125
```
5. Get pods and notice the Starter has ImagePullBackOff
```bash
k6-apache-stress-1-5gfk5                          1/1     Running            0          16m
k6-apache-stress-2-52scp                          1/1     Running            0          16m
k6-apache-stress-starter-cxfsd                    0/1     ImagePullBackOff   0          16m
k6-operator-controller-manager-559fccc44d-nrx2j   2/2     Running            0          26m
```
Pod Events
```bash
  Normal   Scheduled  18m                  default-scheduler  Successfully assigned chaos/k6-apache-stress-starter-cxfsd to ip-10-1-4-142.us-east-2.compute.internal
  Warning  Failed     16m (x6 over 18m)    kubelet            Error: ImagePullBackOff
  Normal   Pulling    16m (x4 over 18m)    kubelet            Pulling image "<repo>/operator:controller-v0.0.7"
  Warning  Failed     16m (x4 over 18m)    kubelet            Failed to pull image "<repo>/operator:controller-v0.0.7": rpc error: code = Unknown desc = failed to pull and unpack image "<repo>/operator:controller-v0.0.7": failed to resolve reference "<repo>/operator:controller-v0.0.7": failed to authorize: failed to fetch anonymous token: unexpected status: 401 Unauthorized
  Warning  Failed     16m (x4 over 18m)    kubelet            Error: ErrImagePull
  Normal   BackOff    3m8s (x66 over 18m)  kubelet            Back-off pulling image "<repo>/operator:controller-v0.0.7"
```